chargebee/src/main/java/com/chargebee/android/billingservice/BillingClientManager.kt

```diff
@@ -253,6 +255,102 @@ class BillingClientManager(context: Context) : PurchasesUpdatedListener {
+                val productDetailsParamsList =
+                    listOf(productDetailsBuilder.build())
+
+                queryAllSubsPurchaseHistory(ProductType.SUBS.value) { subscriptionHistory ->
+                    val oldPurchaseToken: String = getOldPurchaseToken(subscriptionHistory, oldProductId)
+
+                    val billingFlowParams =
+                        BillingFlowParams.newBuilder()
+                            .setProductDetailsParamsList(productDetailsParamsList)
+                            .setSubscriptionUpdateParams(
+                                BillingFlowParams.SubscriptionUpdateParams.newBuilder()
+                                    .setOldPurchaseToken(oldPurchaseToken)
+                                    .setReplaceProrationMode(prorationMode).build()
+                            ).build()
+
+                    billingClient?.launchBillingFlow(mContext as Activity, billingFlowParams)
+                        .takeIf { billingResult ->
+                            billingResult?.responseCode != OK
+                        }?.let { billingResult ->
+                            Log.e(TAG, "Failed to launch billing flow $billingResult")
+                            val billingError = CBException(
+                                ErrorDetail(
+                                    message = GPErrorCode.LaunchBillingFlowError.errorMsg,
+                                    httpStatusCode = billingResult.responseCode
+                                )
+                            )
+                            if (ProductType.SUBS == this.purchaseProductParams.product.type) {
+                                purchaseCallBack?.onError(
+                                    billingError
+                                )
+                            } else {
+                                oneTimePurchaseCallback?.onError(
+                                    billingError
+                                )
+                            }
+                        }
+                }
+            } else {
+                Log.e(TAG, "Failed to fetch product :" + billingResult.responseCode)
+                if (ProductType.SUBS == this.purchaseProductParams.product.type) {
+                    purchaseCallBack?.onError(throwCBException(billingResult))
+                } else {
+                    oneTimePurchaseCallback?.onError(throwCBException(billingResult))
+                }
+
+            }
+        }
+
+    }
+
+    private fun getOldPurchaseToken(subscriptionHistory: List<PurchaseTransaction>?, oldProductId: String): String {
+        val purchaseTransactionHistory = mutableListOf<PurchaseTransaction>()
+        purchaseTransactionHistory.addAll(subscriptionHistory ?: emptyList())
+        val prevProduct: PurchaseTransaction? = purchaseTransactionHistory.find { it.productId.first() == oldProductId }
+        return prevProduct?.purchaseToken ?: ""
```

Comment:
nit: return an empty optional instead of empty string

Reply:
Done